### ISSUE TITLE
[Bug] SLCAND not working inside container with standard network

### ISSUE THREAD

**Issue by JanJamaszyk-FlyNow** [docker/for-linux#1184]
* [x] This is a bug report
* [ ] This is a feature request
* [x] I searched existing issues before opening this one

### Expected behavior
When container is started with `--net=host`:
```
sudo slcand -o -s6 -t hw -S 115200 -F /dev/slcan2usb slcan1
[6] starting on TTY device /dev/slcan2usb
[5] attached TTY /dev/slcan2usb to netdevice slcan0

 

[5] netdevice slcan0 renamed to slcan1
```

### Actual behavior
When container is started **without** `--net=host`:
```
sudo slcand -o -s6 -t hw -S 115200 -F /dev/slcan2usb slcan1
[6] starting on TTY device /dev/slcan2usb
[5] attached TTY /dev/slcan2usb to netdevice slcan0

 

[5] netdevice slcan0 rename to slcan1 failed

 

ioctl SIOCSIFNAME rename: No such device
```

If we look into the source-code of [slcand](https://github.com/linux-can/can-utils/blob/eb66451df280f95a9a12e78b151b8d867e1b78ed/slcand.c#L376) it seems as if `ioctl(fd, SIOCGIFNAME, ifr.ifr_name)` returns a value >= 0 while not actually creating the device. Hence the log-message: `ioctl SIOCSIFNAME rename: No such device` is triggered by `ioctl(s, SIOCSIFNAME, &ifr) < 0` from [here](https://github.com/linux-can/can-utils/blob/eb66451df280f95a9a12e78b151b8d867e1b78ed/slcand.c#L394).

### Steps to reproduce the behavior

1. Connect an SLCAN compatible USB to CAN interface
2. Use Ubuntu 20.04 Base Image
3. Map the serial device into the docker container
4. Start slcand as described above

**Output of `docker version`:**

```
docker version
Client: Docker Engine - Community
 Version:           20.10.2
 API version:       1.41
 Go version:        go1.13.15
 Git commit:        2291f61
 Built:             Mon Dec 28 16:17:43 2020
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

 

Server: Docker Engine - Community
 Engine:
  Version:          20.10.2
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.13.15
  Git commit:       8891c58
  Built:            Mon Dec 28 16:15:19 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.4.3
  GitCommit:        269548fa27e0089a8b8278fc4fc781d7f65a939b
 runc:
  Version:          1.0.0-rc92
  GitCommit:        ff819c7e9184c13b7c2607fe6c30ae19403a7aff
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

**Output of `docker info`:**

```
docker info
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Build with BuildKit (Docker Inc., v0.5.1-docker)

 

Server:
 Containers: 3
  Running: 0
  Paused: 0
  Stopped: 3
 Images: 1044
 Server Version: 20.10.2
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runtime.v1.linux nvidia runc io.containerd.runc.v2
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 269548fa27e0089a8b8278fc4fc781d7f65a939b
 runc version: ff819c7e9184c13b7c2607fe6c30ae19403a7aff
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: default
 Kernel Version: 5.8.0-36-generic
 Operating System: Ubuntu 20.04.1 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 16
 Total Memory: 31.23GiB
 Name: Flyn
 ID: WLLG:E3F3:24ZK:VEIZ:ZYPB:467N:QACM:4CNU:QALK:L6D4:LJDF:YYGT
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
 Default Address Pools:
   Base: 182.240.0.0/16, Size: 24

 

WARNING: No blkio weight support
WARNING: No blkio weight_device support
```

**Comment by JustinOng** [comment]
I've observed that the network interface is actually visible on the host instead of inside the Docker container when started without `--net=host`.

**Comment by JanJamaszyk-FlyNow** [comment]
Wow, good catch Justin!
If I understand this correctly this means that we found a security issue where the container can influence the host environment!

**Comment by JustinOng** [comment]
I don't think this is a security issue. My previous observation was made on
a container started with `--privileged` and `-v /dev:/dev`.

Passing only `--device /dev/ttyUSBS0`, I get:
```
***@***.***:/# slcand -s8 -S 115200 /dev/ttyUSB0 -F
[6] starting on TTY device /dev/ttyUSB0
ioctl TIOCSETD: Operation not permitted
```

**Comment by levivk** [comment]
I can still reproduce this.

Using --device, I get a privilege error when setting the line discipline
```
> docker run -it --device=/dev/ttyACM0 debian bash
> apt update
> apt install can-utils
> slcand -S 3000000 /dev/ttyACM0
# ioctl TIOCSETD: Operation not permitted
```

Using --privileged, slcand succeeds except **_the network device is created on the host and not in the container_**.
```
> docker run -it --privileged debian bash
> apt update
> apt install can-utils iproute2
> slcand -S 3000000 /dev/ttyACM0
> ip link
## No can interface
```
On the host machine with the privileged container still running
```
> ip link
## Shows can0 interface
> sudo ip link set up can0
> candump can0
## Outputs can data
```
Note that I ensured can0 on the host machine did not exist before running the privileged container and stopped existing once the container was exited.

Passing --net=host gives the container access to the can0 interface it created, as would be expected.

Output of `docker version`
```
Client: Docker Engine - Community
 Version:           24.0.7
 API version:       1.43
 Go version:        go1.20.10
 Git commit:        afdd53b
 Built:             Thu Oct 26 09:08:02 2023
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          24.0.7
  API version:      1.43 (minimum version 1.12)
  Go version:       go1.20.10
  Git commit:       311b9ff
  Built:            Thu Oct 26 09:08:02 2023
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.24
  GitCommit:        61f9fd88f79f081d64d6fa3bb1a0dc71ec870523
 runc:
  Version:          1.1.9
  GitCommit:        v1.1.9-0-gccaecfc
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

**Comment by levivk** [comment]
Interestingly, pppd creates a network interface inside of a privileged container just fine.

```
> docker run -it --privileged debian bash
> apt update
> apt install ppp iproute2
> pppd /dev/ttyACM0 noauth
> ip link
## shows interface ppp0
```

**Comment by adrigmz00** [comment]
Any news regarding this topic? I'm currently encountering the same issue and I don't know how to address it.

**Comment by akerouanton** [comment]
> Wow, good catch Justin!
> If I understand this correctly this means that we found a security issue where the container can influence the host environment!

It seems slcand attaches to existing TTY devices to spawn CAN network devices. TTY devices aren't namespaced so this is an expected behavior, and not a security bug. (We can't do better than what the Linux kernel gives us.)

> Passing only `--device /dev/ttyUSBS0`, I get:
> ```
> ***@***.***:/# slcand -s8 -S 115200 /dev/ttyUSB0 -F
> [6] starting on TTY device /dev/ttyUSB0
> ioctl TIOCSETD: Operation not permitted
> ```

Looking at [slcand source code](https://github.com/linux-can/can-utils/blob/eb66451df280f95a9a12e78b151b8d867e1b78ed/slcand.c#L370), you can see that it calls `ioctl(fd, TIOCSETD, N_SLCAN)` which is a privileged operation (when `N_SLCAN` is specified) requiring `CAP_NET_ADMIN` (see [here](https://github.com/torvalds/linux/blob/ac3fd01e4c1efce8f2c054cdeb2ddd2fc0fb150d/drivers/net/can/slcan/slcan-core.c#L823-L824)).

I'm not deeply familiar with CAN hardware, but I think there are more modern ways to communicate with CAN devices than slcand. However, if you must use slcand, you need to run your container with `--cap-add NET_ADMIN` (or `--privileged`, though this gives more privileges to your container).

~~Since slcand works on TTY devices, you must pass the TTY through using `--device /dev/<device-name>`. You don’t need --net=host; the CAN interface will appear in the container’s own network namespace once created.~~

Since the slcan kernel module creates network interfaces in the host netns, you need to run with `--net=host`.

I’ll close this ticket since this is expected kernel behavior and not something Docker can fix at the Engine level. Feel free to keep discussing if you hit other issues.

**Comment by levivk** [comment]
> You don’t need --net=host; the CAN interface will appear in the container’s own network namespace once created.

@akerouanton I think you misunderstand the issue. With slcand, the can network interface is created on the host, not in the container.

**Comment by levivk** [comment]
> Any news regarding this topic? I'm currently encountering the same issue and I don't know how to address it.

Because the CAN network interface is created on the host and not in the container, you can work around this by enabling host networking (i.e. --net=host). This allows the container running slcand to access the can network interface with the side effect of no network isolation.

**Comment by levivk** [comment]
@akerouanton To be clear, I think there _is_ a bug here, but it is not related to TTYs or privilege capabilities. The bug is that, from within a docker container, an interface is created on the host's network namespace. Here is a script to recreate the issue.

```
sudo docker run -i --cap-add=NET_ADMIN debian:13.2-slim bash -s <<- EOF
	apt-get update
	DEBIAN_FRONTEND=noninteractive apt-get install -y can-utils socat
	socat pty,link=/dev/faketty,raw /dev/null,ignoreeof &
	sleep 1
	slcand -F /dev/faketty &
	echo 'Running slcand, now look at host network interfaces'
	sleep infinity
EOF
```

Once the above prints "Running slcand...", inspect the **host** network interfaces and you will find a `can0` interface.

**Comment by akerouanton** [comment]
> I think you misunderstand the issue. With slcand, the can network interface is created on the host, not in the container.

Sorry for the confusion. Yes, you need to run with `--net=host`, or write a Docker networking plugin to run slcand in the host netns and move the interface into a container. (Or write a bespoke script to do something similar.)

> To be clear, I think there _is_ a bug here, but it is not related to TTYs or privilege capabilities. The bug is that, from within a docker container, an interface is created on the host's network namespace.

I don't know if that's a bug, but I know that at least it doesn't come from Docker.

When `ioctl(fd, TIOCSETD, N_SLCAN)` is called by slcand, the kernel calls `slcan_open` (implemented [here](https://github.com/torvalds/linux/blob/ac3fd01e4c1efce8f2c054cdeb2ddd2fc0fb150d/drivers/net/can/slcan/slcan-core.c#L817)) which calls `alloc_candev` to allocate a new `struct net_device *`. This C macro calls `alloc_candev_mqs` (implemented [here](https://github.com/torvalds/linux/blob/ac3fd01e4c1efce8f2c054cdeb2ddd2fc0fb150d/drivers/net/can/dev/dev.c#L288)) which in turn calls `alloc_netdev_mqs` (implemented [here](https://github.com/torvalds/linux/blob/ac3fd01e4c1efce8f2c054cdeb2ddd2fc0fb150d/net/core/dev.c#L11877)). That function itself calls `dev_net_set(dev, &init_net)` (see [here](https://github.com/torvalds/linux/blob/ac3fd01e4c1efce8f2c054cdeb2ddd2fc0fb150d/net/core/dev.c#L11923)) which is responsible for setting the network namespace where the net device lives. Note that `init_net` is the default / host network namespace. Then, in the rest of `alloc_netdev_mqs`, `alloc_candev_mqs` and `slcan_open`, nothing sets the device netns.

If you believe it's a bug, then you should report that on the kernel mailing list.

**Comment by levivk** [comment]
Wow, thanks for the detailed explanation! I looked at the slcan source before, but couldn't find if/how the network namespace was determined. That makes sense now.

By comparison, pppd creates an interface within the container network namespace. I see that after calling `alloc_netdev`, it calls `dev_net_set` ([link](https://github.com/torvalds/linux/blob/30f09200cc4aefbd8385b01e41bde2e4565a6f0e/drivers/net/ppp/ppp_generic.c#L3369)) with the namespace from the current task struct ([link](https://github.com/torvalds/linux/blob/30f09200cc4aefbd8385b01e41bde2e4565a6f0e/drivers/net/ppp/ppp_generic.c#L738)).

So I guess I misunderstood what isolation docker provides and that it's actually up to the driver what namespace a network device goes in, instead of it somehow being determined by docker. Thanks again.

**Comment by akerouanton** [comment]
> So I guess I misunderstood what isolation docker provides and that it's actually up to the driver what namespace a network device goes in, instead of it somehow being determined by docker. Thanks again.

Yes, it's up to the each network driver to assign the netdevice to the current task's netns. AFAIK there's not much netdrivers that don't do it as of now — slcan is an exception.

Since I don't know much about that driver and its use case, did you consider another kind of CAN driver? I believe there are other ones available, but maybe you're limited by what your hardware provides?

Maybe something like https://github.com/moby/moby/issues/50251 could help here. With that, I think you'd be able to automatically start slcand on the host with some udev rules and then use whatever UX we provide (e.g. `--network=netdev=slcan0`) to move that interface into your container. Unfortunately we've no plans to implement that currently (but contributions welcomed!).

**Comment by levivk** [comment]
> Since I don't know much about that driver and its use case, did you consider another kind of CAN driver? I believe there are other ones available, but maybe you're limited by what your hardware provides?

Yeah, the general limitation is devices that connect over USB or Ethernet and can easily provide a SocketCAN interface. Devices implementing the SLCAN protocol seemed to be most readily available. For example, here are two I've used. The first one implements SLCAN over a raw TCP connection, and the second one over a USB virtual serial port.

https://www.coolgear.com/product/1-port-ethernet-to-can-bus-adapter-w-16kv-esd-surge-protection#1
https://www.dfrobot.com/product-2690.html
